OpenRA.Game/Server/Server.cs

```diff
@@ -306,7 +306,7 @@ void ValidateClient(Connection newConn, string data)
 					Slot = LobbyInfo.FirstEmptySlot(),
 					PreferredColor = handshake.Client.Color,
 					Color = handshake.Client.Color,
-					Faction = "Random",
+					Faction = handshake.Client.Faction,
```

Comment:
The `SyncClientToPlayerReference` immediately below already handles the locked faction case, so we only need to worry about maliciously set state.

If you follow the code flow a bit further down still you'll see that it notifies any ServerTraits that implement `IClientJoined`.  One of these is `LobbyCommands`, which has a check to validate the player color inside its `ClientJoined` method.  The best place for the selectable factions check is in there.

You can get a list of all the valid factions with
```csharp
var factions = server.Map.Rules.Actors["world"].TraitInfos<FactionInfo>()
	.Where(f => f.Selectable).Select(f => f.InternalName);
```
and then do something like
```csharp
if (!factions.Contains(client.Faction))
	client.Faction = "Random";
```